### DISCUSSION TITLE
Stumpy & Clustering

### DISCUSSION THREAD

**Discussion by Jalanjii** [TDAmeritrade/stumpy#608] [General]
Hi,

I am interested in clustering with MP measure. I can't find anything on that with Stumpy, is that true or I have missed something?
If you know any other resources please if you can share them.

**Comment by seanlaw** [comment]
@Jalanjii Welcome to the STUMPY community and thank you for your question. So, STUMPY does not provide time series clustering out-of-the-box but based on the [Matrix Profile XII](https://ueaeprints.uea.ac.uk/id/eprint/69244/1/ICDM_2018_Shaghayegh.pdf) paper on MPdist, you should be able to create a pairwise distance matrix between all of your time series using [stumpy.mpdist](https://stumpy.readthedocs.io/en/latest/api.html#stumpy.mpdist):

```
import numpy as np
import stumpy
import hdbscan


n = 1000
Ts = [np.random.rand(n) for i in range(80)]  # Generate 80 random time series
for i in range(20):
    # Generate 20 time series that look similar to `Ts[0]` but with added noise and a roll
    noise = np.random.uniform(0.0, 0.1, size=n)
    Ts.append(np.roll(Ts[0], i * 100) + noise)

# Compute the pairwise distances
distance_matrix = np.zeros((len(Ts), len(Ts)))
m = 50
for i in range(len(Ts)):
    for j in range(i+1, len(Ts)):
        distance = stumpy.mpdist(Ts[i], Ts[j], m)
        distance_matrix[i, j] = distance
        distance_matrix[j, i] = distance
```

And then you can feed this into your preferred clustering algorithm that can accept a distance matrix as input like `hdbscan`:

```
clusterer = hdbscan.HDBSCAN(metric='precomputed', min_cluster_size=3)
clusterer.fit(distance_matrix)
clusterer.labels_
```

**Reply by Jalanjii** [reply]
Thank you. Which algorithm stumpy.mpdist employs? Why you didn't suggest stumpy.scrump for example? I mean, I want to know which package uses the fastest version of matrix profile computation.

**Reply by seanlaw** [reply]
> Thank you. Which algorithm stumpy.mpdist employs? 

It uses the `stump` algorithm under the hood. I strongly recommend reading the [Matrix Profile XII](https://ueaeprints.uea.ac.uk/id/eprint/69244/1/ICDM_2018_Shaghayegh.pdf) paper

> Why you didn't suggest stumpy.scrump for example? 

Your original question asked about clustering time series and not about computing the matrix profile. I recommend going through all of our tutorials in order to gain a better understanding of when you should use the various functions. Note that the matrix profile itself is NOT a distance. Instead it is a vector of distances. 

> I mean, I want to know which package uses the fastest version of matrix profile computation.

For matrix profile computations, please use `stumpy.stump`

**Reply by Jalanjii** [reply]
Thank you, are you intending to integrate scrimp++ or scamp anytime soon in your package?

**Reply by seanlaw** [reply]
`scrimp++`, which is an approximate matrix profile is already implemented as `stumpy.scrump`. What do you mean when you say `scamp`? STUMPY offers a GPU version of `stumpy.stump` called `stumpy.gpu_stump` as well as a multi-server distributed version called `stumpy.stumped`.

What are you really looking for? What are you seeking to accomplish? What problem are you trying to solve?

**Reply by Jalanjii** [reply]
Thanks Sean, I see. I just wanted to have the GPU capability for clustering a large dataset with matrix profile distance. I was quite in a rush that I asked first before reading the tutorials. It seems everything is well included in your package. Thank you for that!

**Reply by seanlaw** [reply]
In case it matters, you may be interested in the GPU version of `stumpy.mpdist` called [stumpy.gpu_mpdist](https://stumpy.readthedocs.io/en/latest/api.html#stumpy.gpu_mpdist). This will use your GPU (only NVIDIA is supported) to compute the underlying matrix profiles

**Reply by Jalanjii** [reply]
That's perfect!

**Reply by Jalanjii** [reply]
Hi Sean, are you intending to do a tutorial to explore the matrix profile method for clustering? Because your tutorials are a gem.

**Reply by seanlaw** [reply]
@Jalanjii Thank you for your kind words. While we plan to eventually publish a tutorial on MPdist, it is unlikely that we'll do anything with clustering time series as the original authors of the matrix profile have already looked at this in [this paper](https://www.cs.ucr.edu/~eamonn/meaningless.pdf)

**Reply by Jalanjii** [reply]
Fair enough. Thank you.

**Reply by Jalanjii** [reply]
@seanlaw If we have multidimensional time series, what would be your take on that for clustering using the distance matrix? For motif discovery and matrix profile, I read about multidimensional motif discovery and MSTUMP. Also, assuming we re not using any data reduction technique.

**Reply by seanlaw** [reply]
Frankly, I don't know. Multidimensional matrix profiles are already hard enough for the average person to understand (it is **not** 1-D matrix profiles stacked on top of each other). So, I wouldn't be able to provide any perspective on what it means to cluster multidimensional time series or how best to accomplish it. Especially since MSTUMP is only available for "self-joins", you can't even do MPdist for multidimensional data.

**Reply by Jalanjii** [reply]
I see. So the best I think would be to do dimensionality reduction and work on the first PCA component or something similar. Thanks

**Reply by Jalanjii** [reply]
@seanlaw It seems there is some work done by someone on multidimensional MP and called it MSTOMP. It was started in this issue https://github.com/matrix-profile-foundation/matrixprofile/issues/15 and finalized here https://github.com/RexKing6/matrixprofile What do you think of that?

MSTOMP: https://github.com/RexKing6/matrixprofile/blob/master/matrixprofile/algorithms/mstomp.py

**Reply by seanlaw** [reply]
@Jalanjii I think there may be some confusion here. MSTOMP was not started by the matrix profile foundation. It was created by the same folks (Eamonn Keogh's group) who created matrix profiles - see this [Matrix Profile VI MSTAMP/MSTOMP paper](https://www.cs.ucr.edu/~eamonn/Motif_Discovery_ICDM.pdf). MSTOMP is exactly what we've implemented and have called it `mstump`. In fact, our implementation is scalable to multiple servers using `dask` and we also have full MDL support and I'm fairly confident that `mstump` will be far more reliable than any other MSTOMP-like implementation that you can find. Is there a reason why you are choosing not to use `mstump`? Having spent a few years perfecting `mstump`, I would not easily trust other implementations because its even trickier to get "right" than 1-D matrix profiles and so I cannot endorse other versions except for `mstump`. I hope that helps.

**Reply by Jalanjii** [reply]
I see. You're absolutely right.

**Reply by Jalanjii** [reply]
> @Jalanjii Welcome to the STUMPY community and thank you for your question. So, STUMPY does not provide time series clustering out-of-the-box but based on the [Matrix Profile XII](https://ueaeprints.uea.ac.uk/id/eprint/69244/1/ICDM_2018_Shaghayegh.pdf) paper on MPdist, you should be able to create a pairwise distance matrix between all of your time series using [stumpy.mpdist](https://stumpy.readthedocs.io/en/latest/api.html#stumpy.mpdist):
> 
> ```
> import numpy as np
> import stumpy
> import hdbscan
> 
> 
> n = 1000
> Ts = [np.random.rand(n) for i in range(80)]  # Generate 80 random time series
> for i in range(20):
>     # Generate 20 time series that look similar to `Ts[0]` but with added noise and a roll
>     noise = np.random.uniform(0.0, 0.1, size=n)
>     Ts.append(np.roll(Ts[0], i * 100) + noise)
> 
> # Compute the pairwise distances
> distance_matrix = np.zeros((len(Ts), len(Ts)))
> m = 50
> for i in range(len(Ts)):
>     for j in range(i+1, len(Ts)):
>         distance = stumpy.mpdist(Ts[i], Ts[j], m)
>         distance_matrix[i, j] = distance
>         distance_matrix[j, i] = distance
> ```
> 
> And then you can feed this into your preferred clustering algorithm that can accept a distance matrix as input like `hdbscan`:
> 
> ```
> clusterer = hdbscan.HDBSCAN(metric='precomputed', min_cluster_size=3)
> clusterer.fit(distance_matrix)
> clusterer.labels_
> ```

@seanlaw Hi Sean, I am back to you with a question about visualizing such time series clusters produced by `hdbscan`, can I use something you recommend?

And given I have a dataframe of this form:
```
date         ts1           ts2           ...
2015         5              6
...
```
For the code you have posted, I have to transpose my dataframe so I can apply it, right? But, my data is multidimensional time series, like the one you created, yet, you stated before that the above code is only for univariate time series as matrix profile distance doesn't support multidimensionality, can you elaborate more please?

**Reply by seanlaw** [reply]
Unfortunately, I have not used `hdbscan` myself so I recommend going through their documentation: https://hdbscan.readthedocs.io/en/latest/index.html

> For the code you have posted, I have to transpose my dataframe so I can apply it, right? But, my data is multidimensional time series, like the one you created, yet, you stated before that the above code is only for univariate time series as matrix profile distance doesn't support multidimensionality, can you elaborate more please?

If you are handling/treating each time series, (i.e., `ts1`), independent of another, (i.e. `ts2`) then you are fundamentally performing a 1-D matrix profile analysis. Whether you are doing a 1-D self-join (`ts1` with `ts1`) or a 1-D AB-join (`ts1` with `ts2`), I would consider both of these to be 1-D since the output matrix profile is 1-D. The only functions in STUMPY that performs multi-dimensional matrix profile analysis is `mstump`/`mstumped`. This function produces multi-dimensional matrix profiles. 

Fundamentally, `mpdist` only produces `1-D` matrix profiles and, hence, it is one-dimensional. Sure, it compares multiple different time series that compares different pairs of time series but the what is computed is a 1-D matrix profile and, hence, `mpdist` is one-dimensional in my head.

**Reply by Jalanjii** [reply]
Thank you. So if I want to perform an analysis on `ts1` only, given this dataframe I have:

```
date           ts1            ts2           ...
1               5              6
2               8              4
...
```
Then I would something like this?

```
import numpy as np
import stumpy
import hdbscan
# Compute the pairwise distances
data = pd.Series(data.ts1)
data_transposed = data.T
Ts = data_transposed 
distance_matrix = np.zeros((len(Ts), len(Ts)))
m = 30
for i in range(len(Ts)):
    for j in range(i+1, len(Ts)):
        distance = stumpy.mpdist(Ts[i], Ts[j], m)
        distance_matrix[i, j] = distance
        distance_matrix[j, i] = distance
clusterer = hdbscan.HDBSCAN(metric='precomputed', min_cluster_size=3)
clusterer.fit(distance_matrix)
clusterer.labels_
```
Is my transpose necessary for the `stumpy.mpdist()` ?
Most importantly, can you interpret the results of your original code, with regard to the cluster labels we got for the 100 time series?

**Reply by seanlaw** [reply]
@Jalanjii I don't understand. You haven't attempted to use `data = pd.Series(data.ts1)` or `data_transposed = data.T` for anything. You've merely defined it. Are you planning to append it to `Ts`? If so then, instead of transposing the Series, I would recommend simply converting the series into a 1-D `numpy` array by doing something like:

```
data = np.asarray(pd.Series(data.ts1))
Ts.append(data)
```

You can check the shape of `data` with `data.shape` to ensure that it is 1-dimensional

**Reply by Jalanjii** [reply]
Sorry, I missed a line, I added `Ts = data_transposed `, I still just want to know about how you would interpret the results you get from this new application, and the one you pasted it in the beginning of this discussion with 100 time series (20 noise ones).

**Reply by seanlaw** [reply]
> Sorry, I missed a line, I added Ts = data_transposed

But then you'd only have a single time series in `Ts` (namely, `data.ts1`). This would be pointless for `mpdist`

> I still just want to know about how you would interpret the results you get from this new application, and the one you pasted it in the beginning of this discussion with 100 time series (20 noise ones)

I'm not sure I understand your question. I would say that "clustering "(and the interpretation of its results) are beyond the scope of support for STUMPY. At the end of the day, `mpdist` takes two time series and computes an approximate "distance measure" between them (please see [Matrix Profile XII paper](https://www.cs.ucr.edu/~eamonn/MPdist_Expanded.pdf)). The code that I've provided above merely computes the set of pairwise "distance measures". Ultimately, it is up to you to interpret the results.

With 100 time series with random noise, there is no "real" interpretation to be made except maybe that the 20 will, hopefully, be clustered together with `Ts[0]` (on which the 20 are based on). However, this is dependent on the clustering algorithm that one chooses and there is no gold standard as you have no way to validate your results. This is the problem with clustering/unsupervised methods. As a gentle reminder, you had stated that you were interested in clustering time series and I had pointed out/warned that:

> it is unlikely that we'll do anything with clustering time series as the original authors of the matrix profile have already looked at this in [this paper](https://www.cs.ucr.edu/~eamonn/meaningless.pdf)

This is because the results are most likely meaningless. Certainly, sometimes, if you are lucky, the results may be useful but, unfortunately, that is beyond the scope of STUMPY support. Our goal is to allow you to compute the matrix profile accurately and faithfully relative to the original papers. The rest is up to you.

**Reply by Jalanjii** [reply]
I see. I mean, all I wanted to know is what does this code do: 

```
for i in range(len(Ts)):
    for j in range(i+1, len(Ts)):
        distance = stumpy.mpdist(Ts[i], Ts[j], m)
        distance_matrix[i, j] = distance
        distance_matrix[j, i] = distance
```
We have your data Ts.shape = (100,1000) (100 time series over a frequency of 1000 if I understand it correctly), and then you fed into `stumpy.mpdist` iteratively, what does distance_matrix store at the end with regard to the 100 time series?

**Reply by seanlaw** [reply]
> We have your data Ts.shape = (100,1000) (100 time series over a frequency of 1000 if I understand it correctly), and then you fed into stumpy.mpdist iteratively, what does distance_matrix store at the end with regard to the 100 time series?

Given that:

```
distance_matrix = np.zeros((len(Ts), len(Ts)))
```

Then the `distance_matrix` is an `n x n` (where `n = len(Ts)`) pairwise distance matrix. Where the `i`th row and `j`the column of `distance_matrix` corresponds to the distance measure computed between `Ts[i]` and `Ts[j]` (i.e., the `i`th time series and the `j`th time series). Note that:

```
for i in range(len(Ts)):
    for j in range(i+1, len(Ts)):
        distance = stumpy.mpdist(Ts[i], Ts[j], m)
        distance_matrix[i, j] = distance
        distance_matrix[j, i] = distance
```

is 100% equivalent to:

```
for i in range(len(Ts)):
    for j in range(len(Ts)):
        distance = stumpy.mpdist(Ts[i], Ts[j], m)
        distance_matrix[i, j] = distance
```

However, the distance matrix is symmetrical along the diagonal (i.e., the diagonal is the "self-distance" between a time series and itself and the upper triangle of distances is the same as the lower triangle) and so we only computed the upper half and then filled in the bottom half due to the built-in redundancy. Otherwise, it would take twice the amount of computational time to compute the upper triangle and then followed by the lower triangle.

**Reply by Jalanjii** [reply]
I see. Thanks, however, I understand this. What I meant was that what does a row or a column mean in this particular matrix computed out of `Ts`  which is `100 x 1000` . I understand what the distance matrix stores when we have a `Ts'` and we build a matrix distance and then the matrix profile based on that like in this tutorial https://stumpy.readthedocs.io/en/latest/Tutorial_The_Matrix_Profile.html. I just want to know how to interpret the distance matrix of your generated data `100 x 1000`. For the linked tutorial, it's beautifully explained there for example, but there, there is only 1 time series.

So for the `100 x 1000` matrix, the first row stores the distances between `ts1` and `ts2` `ts3` and so on.I mean, the subsequences in the above tutorial are now `ts2` `ts3`, right? Then, what does `m` means there, is it going to let us only compare a chunk of time series, basically as if we did `distance = stumpy.mpdist(Ts[i][:m], Ts[j][:m])` supposing `stumpy.mpdist` had a configuration like I just did (I know it's wrong)?

**Reply by seanlaw** [reply]
Have you had a chance to read the [Matrix Profile XII paper](https://www.cs.ucr.edu/~eamonn/MPdist_Expanded.pdf)? This paper should clarify what an "mpdist" is (i.e., it is **not** simply the "matrix profile" between `Ts[i]` and `Ts[j]`).

**Reply by Jalanjii** [reply]
Indeed. I was actually going through it. Thank you for your replies.

**Reply by Jalanjii** [reply]
Hi Sean, I have a single time series, compared to what I had before. I want to build the Distance Matrix as you did in this tutorial, https://stumpy.readthedocs.io/en/latest/Tutorial_The_Matrix_Profile.html, there is a manual way to do that as it is explained, but the Stumpy way I don't find it as far as my reading has gone. For `mpdist`, it is useless I think in this case, right?

**Reply by seanlaw** [reply]
> Hi Sean, I have a single time series, compared to what I had before. I want to build the Distance Matrix as you did in this tutorial, https://stumpy.readthedocs.io/en/latest/Tutorial_The_Matrix_Profile.html, there is a manual way to do that as it is explained, but the Stumpy way I don't find it as far as my reading has gone. For `mpdist`, it is useless I think in this case, right?

@Jalanjii If this is unrelated to clustering multiple time series, can you please start a new discussion?

**Reply by Jalanjii** [reply]
Ok,  I just did. https://github.com/TDAmeritrade/stumpy/discussions/651

**Comment by JaKasb** [comment]
The following paper contains a clustering method.
https://link.springer.com/article/10.1007/s10618-019-00668-6
However, I could not reproduce the clustering method.

**Reply by Jalanjii** [reply]
Thanks!
